File: zZ/cpp/3sum-smaller.cpp
// 259. 3Sum Smaller
// Difficulty : Medium 

// Given an array of n integers nums and a target, find the number of index triplets i, j, k with 0 <= i < j < k < n 
// that satisfy the condition nums[i] + nums[j] + nums[k] < target.

// For example, given nums = [-2, 0, 1, 3], and target = 2.

// Return 2. Because there are two triplets which sums are less than 2:

// [-2, 0, 1]
// [-2, 0, 3]
// Follow up:
// Could you solve it in O(n2) runtime?

// Time:  O(n^2)
// Space: O(1)

class Solution {
public:
  int threeSumSmaller(vector<int>& nums, int target) {
    sort(nums.begin(), nums.end());

    int count = 0;
    for (int i = 0; i < nums.size() - 2; ++i) {
      for (int left = i + 1, right = nums.size() - 1; left < right;) {
        if (nums[i] + nums[left] + nums[right] >= target) {
          --right;
        } else {
          count += right - left;
          ++left;
        }
      }
    }
    return count;
  }
};

File: data-structure-and-algorithm/typescript/leetcode/src/3-sum-smaller.ts
/**
Given an array of n integers nums and a target, find the number of index triplets i, j, k with 0 <= i < j < k < n that satisfy the condition nums[i] + nums[j] + nums[k] < target.

Example:

Input: nums = [-2,0,1,3], and target = 2
Output: 2 
Explanation: Because there are two triplets which sums are less than 2:
             [-2,0,1]
             [-2,0,3]
*/

import { runTestCaseList } from './util/test';

function threeSumSmaller(nums: number[], target: number): number {
    nums.sort((a, b) => a - b);

    let result = 0;

    for (let i = 0; i < nums.length - 2; ++i) {
        for (let left = i + 1, right = nums.length - 1; left < right;) {
            const sum = nums[i] + nums[left] + nums[right];
            if (sum < target) {
                result += right - left;
                ++left;
            } else {
                --right;
            }
        }
    }

    return result;
}

const testInputListCollection = [
    [[-2, 0, 1, 3], 2]
];

const expectedResultList = [
    2
];

runTestCaseList(testInputListCollection, expectedResultList, threeSumSmaller);